crates/ruff/src/rules/flake8_todos/rules.rs

```diff
@@ -223,27 +223,70 @@ impl Violation for MissingSpaceAfterTodoColon {
     }
 }
 
-static TODO_REGEX_SET: Lazy<RegexSet> = Lazy::new(|| {
-    RegexSet::new([
-        r#"^#\s*(?i)(TODO).*$"#,
-        r#"^#\s*(?i)(FIXME).*$"#,
-        r#"^#\s*(?i)(XXX).*$"#,
-    ])
-    .unwrap()
-});
+enum Directive {
+    Todo,
+    Fixme,
+    Xxx,
+}
 
-// Maps the index of a particular Regex (specified by its index in the above TODO_REGEX_SET slice)
-// to the length of the tag that we're trying to capture.
-static PATTERN_TAG_LENGTH: &[usize; 3] = &["TODO".len(), "FIXME".len(), "XXX".len()];
+impl Directive {
+    /// Extract a [`Directive`] from a comment.
+    ///
+    /// Returns the offset of the directive within the comment, and the matching directive tag.
+    fn from_comment(comment: &str) -> Option<(usize, Directive)> {
+        let mut chars = comment.chars().peekable();
+        let mut offset = 0;
+
+        // Detect the leading `#`.
+        if chars.next() == Some('#') {
+            offset += 1;
+        } else {
+            return None;
+        }
 
-static ISSUE_LINK_REGEX_SET: Lazy<RegexSet> = Lazy::new(|| {
-    let patterns: [&str; 3] = [
-        r#"^#\s*(http|https)://.*"#, // issue link
-        r#"^#\s*\d+$"#,              // issue code - like "003"
-        r#"^#\s*[A-Z]{1,6}\-?\d+$"#, // issue code - like "TD003" or "TD-003"
-    ];
-    RegexSet::new(patterns).unwrap()
-});
+        // Skip any whitespace after the leading `#`.
+        while let Some(c) = chars.peek() {
+            if !c.is_whitespace() {
+                break;
+            }
+            offset += c.len_utf8();
+            chars.next();
+        }
```

Comment:
Nit: I wonder if it would be easier to use string operations instead

```
let trimmed = comment.trim_start_matches('#').trim_start();
let offset = comment.text_len() - trimmed.text_len();

// rest
...
```

Reply:
https://github.com/charliermarsh/ruff/pull/4426